src/safeposix/syscalls/net_calls.rs

```diff
@@ -551,148 +619,177 @@ impl Cage {
+                    //If we have already read from a peek but have failed to read more, exit!
+                    if buflen != buflenleft {
+                        return (buflen - buflenleft) as i32;
+                    }
 
-                                if retval < 0 {
-                                    match Errno::from_discriminant(interface::get_errno()) {
-                                        Ok(i) => {
-                                            if sockfdobj.flags & O_NONBLOCK == 0 && i == Errno::EAGAIN {
-                                                if self.cancelstatus.load(interface::RustAtomicOrdering::Relaxed) {
-                                                    // if the cancel status is set in the cage, we trap around a cancel point
-                                                    // until the individual thread is signaled to cancel itself
-                                                    loop { interface::cancelpoint(self.cageid); }
-                                                }
-                                                drop(sockhandle); // release sockhandle temporarily
-                                                sockhandle = sock_tmp.write();
-                                                continue; //received EAGAIN on blocking socket, try again
-                                            }
-                                            return syscall_error(i, "recvfrom", "Internal call to recvfrom failed");
-                                        },
-                                        Err(()) => panic!("Unknown errno value from socket recvfrom returned!"),
-                                    };
-                                    
-                                } else {
-                                    return retval; // we can proceed
+                    match Errno::from_discriminant(interface::get_errno()) {
+                        Ok(i) => {
+                            //We have the recieve timeout set to every one second, so
+                            //if our blocking socket ever returns EAGAIN, it must be
+                            //the case that this recv timeout was exceeded, and we
+                            //should thus not treat this as a failure in our emulated
+                            //socket; see comment in Socket::new in interface/comm.rs
+                            if sockfdobj.flags & O_NONBLOCK == 0 && i == Errno::EAGAIN {
+                                if self.cancelstatus.load(interface::RustAtomicOrdering::Relaxed) {
+                                    // if the cancel status is set in the cage, we trap around a cancel point
+                                    // until the individual thread is signaled to cancel itself
+                                    loop { interface::cancelpoint(self.cageid); }
                                 }
+                                interface::RustLockWriteGuard::<SocketHandle>::bump(sockhandle);
```

Comment:
double check this for potential deadlock if we keep getting EAGAIN

Reply:
its a blocking socket so we should loop continuously. were just hijacking EAGAIN here to check for cancellation